cpp/ql/src/Critical/UseAfterFree.ql

```diff
@@ -11,56 +11,102 @@
  */
 
 import cpp
-import semmle.code.cpp.controlflow.StackVariableReachability
+import semmle.code.cpp.dataflow.new.DataFlow
 
-/** `e` is an expression that frees the memory pointed to by `v`. */
-predicate isFreeExpr(Expr e, StackVariable v) {
-  exists(VariableAccess va | va.getTarget() = v |
-    exists(FunctionCall fc | fc = e |
-      fc.getTarget().hasGlobalOrStdName("free") and
-      va = fc.getArgument(0)
-    )
-    or
-    e.(DeleteExpr).getExpr() = va
-    or
-    e.(DeleteArrayExpr).getExpr() = va
+predicate freeExpr(DataFlow::Node dfe, Expr e) {
+  e = any(DeallocationExpr de).getFreedExpr() and
+  e = dfe.asExpr() and
+  // Ignore any function named realloc
+  exists(DeallocationFunction df | df.getACallToThisFunction().getAChild() = e |
+    not df.getName().regexpMatch(".*realloc") and
+    not df.hasName("MmFreePagesFromMdl") and
+    // Ignore free on pointer dereferences
+    not e = any(PointerDereferenceExpr pde)
   )
 }
 
-/** `e` is an expression that (may) dereference `v`. */
-predicate isDerefExpr(Expr e, StackVariable v) {
-  v.getAnAccess() = e and dereferenced(e)
-  or
-  isDerefByCallExpr(_, _, e, v)
+predicate useExpr(DataFlow::Node dfe, Expr e) {
+  e = dfe.asExpr() and
+  (
+    e = any(ReturnStmt rs).getExpr() or
+    e = any(FunctionCall fc).getAChild() or
+    e = any(PointerDereferenceExpr pde).getOperand() or
+    e = any(PointerFieldAccess pfa).getQualifier() or
+    e = any(ArrayExpr ae).getAChild()
+  ) and
+  not freeExpr(dfe, e)
 }
 
 /**
- * `va` is passed by value as (part of) the `i`th argument in
- * call `c`. The target function is either a library function
- * or a source code function that dereferences the relevant
- * parameter.
+ * Holds if `n` is a dataflow node that is reachable from the
+ * argument of a `DeallocationExpr` that is free'ing the
+ * expression `e`.
  */
-predicate isDerefByCallExpr(Call c, int i, VariableAccess va, StackVariable v) {
-  v.getAnAccess() = va and
-  va = c.getAnArgumentSubExpr(i) and
-  not c.passesByReference(i, va) and
-  (c.getTarget().hasEntryPoint() implies isDerefExpr(_, c.getTarget().getParameter(i)))
+predicate flowsFrom(DataFlow::Node n, Expr e) {
+  freeExpr(n, e)
+  or
+  exists(DataFlow::Node prev |
+    flowsFrom(prev, e) and
+    DataFlow::localFlowStep(prev, n) and
+    not sanitizes(n)
+  )
 }
 
-class UseAfterFreeReachability extends StackVariableReachability {
-  UseAfterFreeReachability() { this = "UseAfterFree" }
-
-  override predicate isSource(ControlFlowNode node, StackVariable v) { isFreeExpr(node, v) }
+/**
+ * Holds if the address of the dataflow node `n` passed to a function call
+ * or if `n` is passed as a references in a function call such that the
+ * memory pointed to by `n` may be changed by the function call.
+ */
+predicate sanitizes(DataFlow::Node n) {
+  n.asIndirectExpr() = any(AddressOfExpr aoe)
+  or
+  exists(Call c | callByReference(c, n.asVariable()))
```

Comment:
I don't think this does what you want. `n.asVariable()` gives you back a `Variable` when `n` is a `VariableNode`, but there's never local flow to/from a `VariableNode`. Instead, you need to do something along these lines:
```suggestion
  exists(Call c | callByReference(c, n.asExpr().(VariableAccess).getTarget())
```
(or possibly `n.asIndirectExpr()`, I'll let you play with that 😄).

Reply:
It needs to be 
```
exists(Call c | callByReference(c, n.asIndirectExpr().(VariableAccess).getTarget())
```
I will fix.